libraries/chain/include/graphene/chain/custom_authorities_utils.hpp

```diff
@@ -0,0 +1,327 @@
+}
+
+template <typename K, typename T>
+inline int64_t to_integer(const map<K, T>& value)
+{
+   return static_cast<int64_t>(value.size());
+}
+
+template <typename K, typename T>
+inline int64_t to_integer(const flat_map<K, T>& value)
+{
+   return static_cast<int64_t>(value.size());
+}
+
+template <typename K, typename T>
+inline int64_t to_integer(const unordered_map<K, T>& value)
+{
+   return static_cast<int64_t>(value.size());
+}
+
+template <typename T>
+void is_type_supported_by_restriction()
+{
+   FC_THROW("Type is not supprted by restruction. Typename = ${name}", ("name", fc::get_typename<T>::name()));
+}
+   
+template <class T>
+bool is_equal(const T& left, const T& right)
+{
+   FC_THROW("Can't compare types. Type '${type_name}' don't support == operator.", ("type_name", fc::get_typename<T>::name()));
+}
+   
+template <typename T>
+const T& get(const generic_member& a_variant)
+{
+   FC_THROW("Can't fetch value. Type '${type_name}' is not supported for now.", ("type_name", fc::get_typename<T>::name()));
+}
+
+#define GRAPHENE_RESTRICTION_TYPE(type) \
+inline bool is_equal(const type& left, const type& right) {   return left == right; } \
+\
+template <> \
+inline const type& get<type>(const generic_member& a_variant) { return a_variant.get<type>(); } \
+\
+template <> inline void is_type_supported_by_restriction<type>() {} \
+
+   GRAPHENE_RESTRICTION_TYPE(uint8_t);
+   GRAPHENE_RESTRICTION_TYPE(uint16_t);
+   GRAPHENE_RESTRICTION_TYPE(uint32_t);
+   GRAPHENE_RESTRICTION_TYPE(asset_id_type);
+   GRAPHENE_RESTRICTION_TYPE(account_id_type);
+   GRAPHENE_RESTRICTION_TYPE(balance_id_type);
+   GRAPHENE_RESTRICTION_TYPE(proposal_id_type);
+   GRAPHENE_RESTRICTION_TYPE(fba_accumulator_id_type);
+   GRAPHENE_RESTRICTION_TYPE(limit_order_id_type);
+   GRAPHENE_RESTRICTION_TYPE(withdraw_permission_id_type);
+   GRAPHENE_RESTRICTION_TYPE(witness_id_type);
+   GRAPHENE_RESTRICTION_TYPE(force_settlement_id_type);
+   GRAPHENE_RESTRICTION_TYPE(committee_member_id_type);
+   GRAPHENE_RESTRICTION_TYPE(public_key_type);
+   GRAPHENE_RESTRICTION_TYPE(time_point_sec);
+   GRAPHENE_RESTRICTION_TYPE(bool);
+   GRAPHENE_RESTRICTION_TYPE(unsigned_int);
+   GRAPHENE_RESTRICTION_TYPE(vector<char>);
+   GRAPHENE_RESTRICTION_TYPE(string);
+   GRAPHENE_RESTRICTION_TYPE(asset);
+   GRAPHENE_RESTRICTION_TYPE(price);
+   GRAPHENE_RESTRICTION_TYPE(price_feed);
+   GRAPHENE_RESTRICTION_TYPE(share_type);
+   GRAPHENE_RESTRICTION_TYPE(vesting_policy_initializer);
+   GRAPHENE_RESTRICTION_TYPE(worker_initializer);
+   GRAPHENE_RESTRICTION_TYPE(extensions_type);
+   GRAPHENE_RESTRICTION_TYPE(future_extensions);
+   GRAPHENE_RESTRICTION_TYPE(vector<predicate>);
+   GRAPHENE_RESTRICTION_TYPE(authority);
+   GRAPHENE_RESTRICTION_TYPE(flat_set<account_id_type>);
+   GRAPHENE_RESTRICTION_TYPE(flat_set<public_key_type>);
+
+#undef GRAPHENE_RESTRICTION_TYPE
+   
+// TODO: should be discussed with bitshares devs and may be fixed in the future
```

Comment:
What does it do and what is it used for?

Reply:
This logic was copied from Abit implementation. 
This is calculation of number of units for restriction (later it is used in fee calculation for operation in which restriction is created or updated)
We've agreed with Stefan to review it and discuss.

Reply:
@abitmore @xeroc @sschiessl-bcp  
Hi, guys 
Could we discuss how fee calculation should be implemented?
At some point we've agreed to postpone discussion  of fee calculation logic to code review. (Till this moment we've agreed just to copy Abit's logic). 

Do we have some view of fee calculation for custom authorities feature? 
Specification describes couple of algorithms but they differs from Abit's scheme.

Reply:
Can you present abit's scheme please?

Reply:
Sure,

On create and update operations fee is calculated with next formula:

```
share_type custom_authority_create_operation::calculate_fee( const fee_parameters_type& k )const
{
   share_type core_fee_required = k.basic_fee;

   if( enabled )
   {
      share_type unit_fee = k.price_per_k_unit;
      unit_fee *= (valid_to - valid_from).to_seconds();
      uint64_t restriction_units = 0;
      for( const auto& r : restrictions )
      {
         units_fetcher visitor;
         restriction_units += r.visit(visitor);
      }
      unit_fee *= restriction_units;
      unit_fee /= 1000;
      core_fee_required += unit_fee;
   }

   return core_fee_required;
}
```

In this line we calculate number of units for each restriction:
`restriction_units += r.visit(visitor);`

Units for the restriction are calculated depending on argument type.

Here is Abits visitor calculating units for argument type:

```
struct argument_get_units_visitor
{
   typedef uint64_t result_type;

   template<typename T>
   inline result_type operator()( const T& t )
   {
      return 1;
   }

   inline result_type operator()( const fc::sha256& t )
   {
      return 4;
   }

   inline result_type operator()( const public_key_type& t )
   {
      return 4;
   }

   inline result_type operator()( const string& t )
   {
      return ( t.size() + 7 ) / 8;
   }

   template<typename T>
   inline result_type operator()( const flat_set<T>& t )
   {
      return t.size() * (*this)( *((const T*)nullptr) );
   }

   template<typename T>
   inline result_type operator()( const flat_set<string>& t )
   {
      result_type result = 0;
      for( const auto& s : t )
      {
         result += ( s.size() + 7 ) / 8;
      }
      return result;
   }

   result_type operator()( const vector<restriction>& t )
   {
      result_type result = 0;
      for( const auto& r : t )
      {
         result += r.argument.visit(*this);
      }
      return result;
   }
};
```

Reply:
I can look in the code :) 

But what is the difference to how you want to do it, or the specs say, in your words?

Reply:
For example create_custom_authority_operation.

Specification says we should either implement:
> A fixed high fee independent of authorities content

Or 
> Tied to the duration and complexity of the custom active authority. Transaction fee is then fee = flat_fee + basic_fee * duration where basic_fee is calculated according to complexity (e.g. size of authority, number of restrictions and etc.). The period duration refers to the time that is still left, not the set duration (i.e. date_to - max(now, date_from))

Abit implemented the second way. But for me specification is not clear on how the calculation formula should look like. 
Abit's formula uses units calculated for each restriction (which was not mentioned in spec) and has some extra actions (not described in spec). For example: 
> unit_fee /= 1000;

So questions are:
1. Should we reuse Abit logic for fee calculation?  Or should we implement some other formula? Or should we just use fixed fee for each operation?
2. If Abit's logic is actual, then we need to chose units number for each possible restriction argument type.
(I'll prepare list of types if we decide to do so)

@sschiessl-bcp  @abitmore

Reply:
Gut feeling says use what abit has laid out for us. Specifications talk about `basic_fee is calculated according to complexity`, and how we calculate the complexity is now up to us.

Reply:
My intention is to make the fee "fair". Fee rate is in Satoshi. Think about use cases, e.g. what's a fair price for a one-month permission granted to a single key so it can be used to place orders.

Reply:
Ok, we should left Abit's implementation. I'll we need just to calculate units for all types.
Am I correct?

(I'll prepare list of types we need to specify units number for)

@abitmore  @sschiessl-bcp

Reply:
So guys, here are types we need to support (those types are present in various operations):
   vote_id_type,
   uint8_t,
   uint16_t,
   uint32_t,
   asset_id_type,
   account_id_type,
   balance_id_type,
   proposal_id_type,
   fba_accumulator_id_type,
   limit_order_id_type,
   withdraw_permission_id_type,
   witness_id_type,
   force_settlement_id_type,
   committee_member_id_type,
   public_key_type,
   time_point_sec,
   bool,
   unsigned_int,
   vector<char>,
   string,
   asset,
   price,
   price_feed,
   share_type,
   vesting_policy_initializer,
   worker_initializer,
   extensions_type,
   future_extensions,
   vector<predicate>,
   authority,
   vesting_balance_id_type,
   flat_set<account_id_type>,
   flat_set<public_key_type>

How will we calculate units for them?

> Think about use cases, e.g. what's a fair price for a one-month permission granted to a single key so it can be used to place orders.
I'm not sure I can decide on those rules and fee prices. 

@sschiessl-bcp  Could you please clearify, am I responsible for this?